packages/SwingSet/src/kernel/kernel.js

```diff
@@ -375,75 +386,164 @@ export default function buildKernel(
    */
   async function deliverAndLogToVat(vatID, kd, vd) {
+    vatRequestedTermination = undefined;
+    illegalSyscall = undefined;
     assert(vatWarehouse.lookup(vatID));
     // Ensure that the vatSlogger is available before clist translation.
     const vs = kernelSlog.provideVatSlogger(vatID).vatSlog;
     try {
       /** @type { VatDeliveryResult } */
       const deliveryResult = await vatWarehouse.deliverToVat(vatID, kd, vd, vs);
-
       insistVatDeliveryResult(deliveryResult);
       // const [ ok, problem, usage ] = deliveryResult;
-      if (deliveryResult[0] === 'ok') {
-        return { metering: deliveryResult[2] };
-      } else {
-        // probably a hard metering fault, or a bug in the vat's dispatch()
-        return { terminate: deliveryResult[1] }; // might be dead
+
+      const status = { illegalSyscall, vatRequestedTermination };
+      // we get metering for all non-throwing deliveries (both 'ok'
+      // and 'error') except for hard metering faults
+      status.metering = deliveryResult[2];
+      if (deliveryResult[0] !== 'ok') {
+        // kernel-panic -worthy delivery problems (unexpected worker
+        // exit, pipe problems) are signalled with an exception, so
+        // non-'ok' status means:
+        //
+        // * hard metering fault (E_TOO_MUCH_COMPUTATION or
+        //   E_STACK_OVERFLOW or E_NOT_ENOUGH_MEMORY), and the worker is
+        //   dead
+        // * a bug in the vat's dispatch() or liveslots, and we
+        //   shouldn't trust the surviving worker
+        // * Liveslots observed a vat-fatal error, like startVat()
+        //   noticed buildRootObject() did not restore all virtual
+        //   Kinds, and we're not going to use the worker
+        //
+        // So in all cases, our caller should abandon the worker.
+        await vatWarehouse.killWorker(vatID);
+        // TODO: does killWorker work if the worker just died?
+        status.deliveryError = deliveryResult[1];
       }
+      return harden(status);
     } catch (e) {
-      // log so we get a stack trace
+      // log so we get a stack trace before we panic the kernel
       console.error(`error in kernel.deliver:`, e);
       throw e;
     }
   }
 
+  /**
+   * Given the results of a delivery, build a set of instructions for
+   * finishing up the crank.
+   *
+   * Two flags influence this:
+   *  `decrementReapCount` is used for deliveries that run userspace code
+   *  `meterID` means we should check a meter
+   *
+   * @param {VatID} vatID
+   * @param {DeliveryStatus} status
+   * @param {boolean} decrementReapCount
+   * @param {MeterID=} meterID
+   * @returns {CrankResults}
+   */
+  function deliveryCrankResults(vatID, status, decrementReapCount, meterID) {
+    let meterUnderrun = false;
+    let computrons;
+    if (status.metering?.compute) {
+      computrons = BigInt(status.metering.compute);
+    }
+    // TODO metering.allocate, some day
+
+    const results = { computrons };
```

Comment:
Nit (and remove explicit `@returns`)

```suggestion
    /** @type {CrankResults} */
    const results = { computrons };
```